[exec] コマンド実行時にexpanderとpath探索を呼び出す

nafuka11 opened the pull request:
Fix #38

## やったこと
exec_command時に `expand_tokens` と `build_cmd_path` を実行するようにしました。leakしないようにfreeしてます。
norm対策でexec_commandの処理を `exec_command.c` に分割しました。

以下のコマンドでleakしないことを確認しています。
```bash
while [ 1 ]
do
leaks minishell
sleep 1
done
```
exitすると、valgrindではenvsとgnlのlineが残っていて definitely lost になります。

## 気になっていること
1. norm対策で関数を分けたのですが、分かりやすい命名かどうか気になっています。
   - `convert_tokens` が convert_argsをwrapしてるのが分かりやすいかどうか。
2. ft_free_set をlibftに追加するか迷っています。
   - freeしつつ指定ポインタをセットする関数です。

nafuka11 committed:
chore(Makefile): SRCS更新

nafuka11 committed:
feat(exec): exec_commandでexpand, path探索を実行するように

nafuka11 committed:
fix(exec): normエラー対応（行数、ファイル分割）

exec_command.cを追加。
exec_command()のtoken展開・args変換をconvert_token()へ。
ポインタをfreeしつつ値をセットするft_free_set()を追加。

nafuka11 committed:
fix(exec): 生成した環境変数文字列配列をfreeするよう修正。関数分割

exec_binary: forkした子プロセスの処理

ToYeah commented on srcs/exec/exec_command.c:
```diff
@@ -0,0 +1,68 @@
+		exec_builtin(args);
+		return ;
+	}
+	ft_free_set(&args[0], build_cmd_path(args[0]));
+	create_pipe(state, new_pipe);
+	if ((pid = fork()) < 0)
+		error_exit();
+	if (pid == 0)
+		exec_binary(args, state, old_pipe, new_pipe);
```
build_cmd_pathがNULLを返した時の処理が必要だと思います！
execveにNULLがわたって`Bad address`が表示されます。
本家は`コマンド名: command not found`の形式でエラーが出るみたいです。

```
minishell > kakakkakakak
Bad address
```

ToYeah replied:
存在しないファイルやディレクトリを指定したときのエラーはexecve側で拾えそうなので、`引数: `の形式でエラー出力する関数があると使いまわし効くかもしれません！
```
$ /bin/ls/ls
bash: /bin/ls/ls: Not a directory
```

nafuka11 replied:
`error_exit` 、 `print_error` でコマンドを指定できるよう修正しました。
execveだけコマンド入れてます。他はNULLにして、コマンド未指定扱いにしています。

`error_exit` はまだ終了ステータスは設定できません。
エラー周りの詳細（コマンドをエラーメッセージに含めるか？終了ステータスは何か？を各エラーで検証）は別Issueにしようかなと思います。

nafuka11 replied:
Bad address も修正し、 command not found になるようにしています。

ToYeah reviewed (changes requested):
`build_cmd_path`の返り値チェックだけ追加して、エラー関連は別issue立てて対応でもいいかもしれません！

nafuka11 committed:
Merge remote-tracking branch 'origin' into 38-call-expander-and-build-path

nafuka11 committed:
fix(exec): Bad address修正。command not foundを表示しstatus127を返すように

nafuka11 committed:
fix: error_exitにcommandを指定できるように

execveの時だけcommandを指定しています。それ以外は未指定です
エラーメッセージの形式
- command指定時 minishell: command: strerror
- 未指定の時 minishell: strerror

nafuka11 committed:
test: absolute_path.txtからsimple_command.txtにリネーム

nafuka11 committed:
test: bashのエラーメッセージをminishell用に置換するように

nafuka11 committed:
test(cases): 存在しないファイル、パスなしコマンドのテスト追加

ToYeah reviewed (approved):
問題ないと思います！！！
